modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/util/ProjectedTuple.java

```diff
@@ -0,0 +1,322 @@
+    }
+
+    @Override
+    public byte byteValue(int col) {
+        return delegate.byteValue(projection[col]);
+    }
+
+    @Override
+    public Byte byteValueBoxed(int col) {
+        return delegate.byteValueBoxed(projection[col]);
+    }
+
+    @Override
+    public short shortValue(int col) {
+        return delegate.shortValue(projection[col]);
+    }
+
+    @Override
+    public Short shortValueBoxed(int col) {
+        return delegate.shortValueBoxed(projection[col]);
+    }
+
+    @Override
+    public int intValue(int col) {
+        return delegate.intValue(projection[col]);
+    }
+
+    @Override
+    public Integer intValueBoxed(int col) {
+        return delegate.intValueBoxed(projection[col]);
+    }
+
+    @Override
+    public long longValue(int col) {
+        return delegate.longValue(projection[col]);
+    }
+
+    @Override
+    public Long longValueBoxed(int col) {
+        return delegate.longValueBoxed(projection[col]);
+    }
+
+    @Override
+    public float floatValue(int col) {
+        return delegate.floatValue(projection[col]);
+    }
+
+    @Override
+    public Float floatValueBoxed(int col) {
+        return delegate.floatValueBoxed(projection[col]);
+    }
+
+    @Override
+    public double doubleValue(int col) {
+        return delegate.doubleValue(projection[col]);
+    }
+
+    @Override
+    public Double doubleValueBoxed(int col) {
+        return delegate.doubleValueBoxed(projection[col]);
+    }
+
+    @Override
+    public BigDecimal decimalValue(int col, int decimalScale) {
+        return delegate.decimalValue(projection[col], decimalScale);
+    }
+
+    @Override
+    public BigInteger numberValue(int col) {
+        return delegate.numberValue(projection[col]);
+    }
+
+    @Override
+    public String stringValue(int col) {
+        return delegate.stringValue(projection[col]);
+    }
+
+    @Override
+    public byte[] bytesValue(int col) {
+        return delegate.bytesValue(projection[col]);
+    }
+
+    @Override
+    public UUID uuidValue(int col) {
+        return delegate.uuidValue(projection[col]);
+    }
+
+    @Override
+    public BitSet bitmaskValue(int col) {
+        return delegate.bitmaskValue(projection[col]);
+    }
+
+    @Override
+    public LocalDate dateValue(int col) {
+        return delegate.dateValue(projection[col]);
+    }
+
+    @Override
+    public LocalTime timeValue(int col) {
+        return delegate.timeValue(projection[col]);
+    }
+
+    @Override
+    public LocalDateTime dateTimeValue(int col) {
+        return delegate.dateTimeValue(projection[col]);
+    }
+
+    @Override
+    public Instant timestampValue(int col) {
+        return delegate.timestampValue(projection[col]);
+    }
+
+    @Override
+    public ByteBuffer byteBuffer() {
+        normalizeIfNeeded();
+
+        return delegate.byteBuffer();
+    }
+
+    private void normalizeIfNeeded() {
+        if (normalized) {
+            return;
+        }
+
+        if (schema != null) {
```

Comment:
I'm not insisting, but it seems that the code may be cleaner if we split this into two classes (a base without a schema, and a descendant with a schema). I don't know what to call them beautifully. , we can add a factory method to create a ProjectedTuple, which will create the desired internal implementation.

Reply:
> but it seems that the code may be cleaner if we split this into two classes

fixed, I guess. I split it onto two different classes though: one abstract and 2 particular implementations.

> we can add a factory method to create a ProjectedTuple, which will create the desired internal implementation

this looks like overkill, imo. There is only one implementation by constructors signature. And I don't think we will have new implementations anytime soon.